### PR TITLE
[Dygraph to static graph]JIT/Trace

### PR DESCRIPTION
**Pull request by sneaxiy**
Add `fluid.dygraph.jit.trace` API to trace the operators run in dygraph mode, so that it can be converted to a `ProgramDesc`.

The usage is:

```Python
import paddle.fluid.dygraph.jit as jit

outs, program = jit.trace(module=module, inputs=[image, label], 
                          feed_names=['image', 'label'], fetch_names=['loss'])
```

The `outs` are the output of `module(*inputs)` and program is the converted program with the given feed names and fetch names. 

**This API is not public yet in this PR because we should discuss the detailed API style later.**

### CODE REVIEW

File: python/paddle/fluid/dygraph/jit.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def extract_vars(inputs):
+    result_list = []
+    _extract_vars(inputs, result_list)
+    return result_list
+
+
+@dygraph_only
+def trace(module, inputs, feed_names=None, fetch_names=None):
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/jit.py:47]: maybe we need some doc and sample code for this method

File: python/paddle/fluid/dygraph/jit.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+def _extract_vars(inputs, result_list):
+    if isinstance(inputs, Variable):
+        result_list.append(inputs._ivar)
+
+    if isinstance(inputs, (list, tuple)):
+        for var in inputs:
+            _extract_vars(var, result_list)
+
+
+def extract_vars(inputs):
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/jit.py:40]: Maybe there is no need for user to access to this method

**Review by JiabinYang — COMMENTED** [review]
some comments

**Review by JiabinYang — APPROVED** [review]
LGTM

File: python/paddle/fluid/dygraph/jit.py (hunk 2 above)
Comment (sneaxiy) [reply python/paddle/fluid/dygraph/jit.py:40]: Yes, we would not expose this method to users.

File: python/paddle/fluid/dygraph/jit.py (hunk 1 above)
Comment (sneaxiy) [reply python/paddle/fluid/dygraph/jit.py:47]: Thanks! I would add more docs and sample codes in next PR!

**Pull request merged**